### PR TITLE
feat(test): moving unit tests over to vitest

### PR DESCRIPTION
**Pull request by erunion**
## 🧰 Changes

I've been wildy impressed with [Vitest](https://npm.im/vitest) so this is moving us off Jest and over to that. The performance gains aren't going to break the bank, but I'll happily take a couple seconds off of the `api` suite any day.

##### api
```
Test Suites: 18 passed, 18 total
Tests:       7 todo, 216 passed, 223 total
Snapshots:   0 total
Time:        29.639 s
Ran all test suites matching /test\/core\/index.test.ts|test\/core\/getJSONSchemaDefaults.test.ts|test\/core\/parseResponse.test.ts|test\/core\/prepareParams.test.ts|test\/core\/prepareAuth.test.ts|test\/auth.test.ts|test\/index.test.ts|test\/cache.test.ts|test\/cli\/codegen\/languages\/typescript\/utils.test.ts|test\/cli\/codegen\/languages\/typescript.test.ts|test\/cli\/storage.test.ts|test\/fetcher.test.ts|test\/cache-tmp.test.ts|test\/dist.test.ts|test\/cache-custom.test.ts|test\/config.test.ts|test\/integration.test.ts|test\/server.test.ts/i.
npx jest --coverage   77.67s user 5.05s system 275% cpu 30.033 total
```

##### httpsnippet-client-api
```
Test Suites: 1 passed, 1 total
Tests:       68 passed, 68 total
Snapshots:   0 total
Time:        4.174 s, estimated 5 s
Ran all test suites.
npx jest  6.69s user 0.40s system 149% cpu 4.739 total
```

##### api
```
 Test Files  18 passed (18)
      Tests  216 passed | 7 todo (223)
   Start at  18:22:46
   Duration  26.88s (transform 817ms, setup 305ms, collect 4.15s, tests 35.72s, environment 2ms, prepare 1.38s)

npx vitest run  56.27s user 4.04s system 220% cpu 27.316 total
```

##### httpsnippet-client-api
```
 Test Files  1 passed (1)
      Tests  68 passed (68)
   Start at  18:24:35
   Duration  3.32s (transform 184ms, setup 0ms, collect 245ms, tests 2.87s, environment 0ms, prepare 52ms)

npx vitest run  5.34s user 0.50s system 155% cpu 3.750 total
```

### CODE REVIEW

File: packages/api/test/.eslintrc [hunk 1]
```diff
@@ -1,6 +1,27 @@
 {
-  "extends": "@readme/eslint-config/testing",
+  "extends": ["plugin:vitest/all"],
```
Comment (erunion) [inline packages/api/test/.eslintrc:2]: Will replace these with `@readme/eslint-config/testing/vitest` when https://github.com/readmeio/standards/pull/723 is merged and published.

File: packages/httpsnippet-client-api/package.json [hunk 2]
```diff
@@ -9,7 +9,7 @@
     "lint:types": "tsc --noEmit",
     "prebuild": "rm -rf dist/",
     "prepack": "npm run build",
-    "test": "jest --coverage"
+    "test": "vitest run --coverage"
```
Comment (kanadgupta) [inline packages/httpsnippet-client-api/package.json:12]: nit but [per their docs](https://vitest.dev/guide/cli.html#vitest), `vitest` auto-runs in watch mode in dev and non-watch in CI. thoughts on this?

```suggestion
    "test": "vitest --coverage"
```

File: packages/httpsnippet-client-api/package.json [hunk 3]
```diff
@@ -36,14 +36,13 @@
...
     "@readme/openapi-parser": "^2.4.0",
     "@types/content-type": "^1.1.5",
-    "@types/jest": "^29.5.2",
     "@types/stringify-object": "^4.0.2",
+    "@vitest/coverage-v8": "^0.34.1",
     "api": "file:../api",
     "fetch-mock": "^9.11.0",
     "isomorphic-fetch": "^3.0.0",
-    "jest": "^29.6.1",
-    "ts-jest": "^29.1.1",
```
Comment (kanadgupta) [inline packages/httpsnippet-client-api/package.json:45]: love how we don't need any additional packages to support TS 😌

File: packages/api/test/helpers/vitest.matchers.ts [hunk 4]
```diff
@@ -52,39 +51,38 @@ expect.extend({
     };
   },
 
-  toHaveCustomUserAgent(this: jest.MatcherUtils, headers: string[]) {
-    const { printReceived } = this.utils;
+  toHaveCustomUserAgent(headers: string[]) {
+    // const { printReceived } = this.utils;
```
Comment (kanadgupta) [inline packages/api/test/helpers/vitest.matchers.ts:55]:
```suggestion
```

File: packages/api/test/global.d.ts [hunk 5]
```diff
@@ -1,3 +1 @@
-import 'jest-extended';
```
Comment (kanadgupta) [inline packages/api/test/global.d.ts:1]: we really were living like gremlins huh

File: packages/api/test/auth.test.ts [hunk 6]
```diff
@@ -192,7 +193,7 @@ describe('#auth()', () => {
       // Because the `POST /anything` operation allows either an OAuth2 token or Basic Auth the
       // quirks case we're testing is that you should be able to supply either a single OAuth2 token
       // or a username+password and it should be able to intelligently handle both.
-      // eslint-disable-next-line jest/no-standalone-expect
+      // eslint-disable-next-line vitest/no-standalone-expect
```
Comment (kanadgupta) [inline packages/api/test/auth.test.ts:196]: FWIW i feel like this could live in its own `it` block

File: packages/api/package.json [hunk 7]
```diff
@@ -13,8 +13,8 @@
     "lint:types": "tsc --noEmit",
     "prebuild": "rm -rf dist/; npm run version",
     "prepack": "npm run build",
-    "test": "jest --coverage $(find test -name '*.test.ts' -not -path '*/smoketest.test.ts')",
-    "test:smoke": "npx jest test/cli/codegen/languages/typescript/smoketest.test.ts",
+    "test": "vitest run --coverage",
+    "test:smoke": "vitest run --config=vitest-smoketest.config.ts ",
```
Comment (kanadgupta) [inline packages/api/package.json:17]: similar to my suggestion in this comment

```suggestion
    "test": "vitest --coverage",
    "test:smoke": "vitest --config=vitest-smoketest.config.ts ",
```

File: packages/api/.npmignore [hunk 8]
```diff
@@ -4,5 +4,5 @@ test/
 .eslint*
 .gitignore
 .prettier*
-jest.config.js
+vitest.*
```
Comment (kanadgupta) [inline packages/api/.npmignore:7]: it needs to be this to pick on up `vitest-smoketest.config.ts` right?

```suggestion
vitest*
```

**Review by kanadgupta — CHANGES_REQUESTED** [review]
looks good for the most part, just a few small comments! very jealous of how easy of a drop-in replacement this was... https://github.com/readmeio/rdme/pull/857 is looking rough 😬

**Pull request merged**